phoenix-core-client/src/main/java/org/apache/phoenix/util/ScanUtil.java

```diff
@@ -1615,14 +1615,21 @@ public static boolean isIncompatibleClientForServerReturnValidRowKey(Scan scan)
     public static PageFilter removePageFilter(Scan scan) {
         Filter filter = scan.getFilter();
         if (filter != null) {
+            PagingFilter pagingFilter = null;
             if (filter instanceof PagingFilter) {
-                filter = ((PagingFilter) filter).getDelegateFilter();
+                pagingFilter = (PagingFilter) filter;
+                filter = pagingFilter.getDelegateFilter();
                 if (filter == null) {
                     return null;
                 }
             }
             if (filter instanceof PageFilter) {
-                scan.setFilter(null);
+                if (pagingFilter != null) {
+                    pagingFilter.setDelegateFilter(null);
+                    scan.setFilter(pagingFilter);
```

Comment:
This helps when filter is of both types: HBase PageFilter and Phoenix PagingFilter (which happens only when LIMIT is present), so that we want to retain PagingFilter for performing page timeouts?

Reply:
Yes, the LIMIT translates to HBase PageFilter and the Phoenix PagingFilter prevents rpc timeouts.